### PR TITLE
Prevent placing an order with a disabled shipping method

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12
| Bug fix?        | yes
| New feature?    | kinda
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | replaces #12870
| License         | MIT                                                          |

Currently, when we disable a shipping method, and someone is after the shipping method choice step and has the disabled shipping method selected, they'll be able to place the order. After merging this PR, the customer is going to see 
> The "%shippingMethodName%" shipping method is not available, please choose another one.
message.

### CODE REVIEW

File: features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature [hunk 1]
```diff
@@ -0,0 +1,20 @@
+@checkout
+Feature: Preventing placing an order with a disabled shipping method
+    In order to ship my order properly
+    As a Customer
+    I want to not be able to place an order with a disabled shipping method
```
Comment (TheMilek) [inline features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature]:
```suggestion
    I should not be able to place an order with a disabled shipping method
```

File: features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Ubi T-Shirt" priced at "$19.99"
+        And the store has "Raven Post" shipping method with "$4.00" fee
+        And the store allows paying "Offline"
+        And I am a logged in customer
+
+    @ui @api
+    Scenario: Prevent placing an order with a shipping method disabled after completing the shipping method choice step
```
Comment (TheMilek) [inline features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature]:
```suggestion
    Scenario: Preventing from placing an order with a shipping method disabled after completing the shipping method choice step
```
🤔

**Review by TheMilek — DISMISSED** [review]

File: features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature (hunk 1 above)
Comment (NoResponseMate) [inline features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature]:
```suggestion
    I want to be prevented from placing an order with a disabled shipping method
```

File: features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature [hunk 3]
```diff
@@ -0,0 +1,20 @@
+@checkout
+Feature: Preventing placing an order with a disabled shipping method
+    In order to ship my order properly
```
Comment (NoResponseMate) [inline features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature]:
```suggestion
    In order to have my order processed smoothly
```
or
```suggestion
    In order to not provide invalid data for my order
```
or
```suggestion
    In order to have my order shipped without issues
```
or something similar, we're in the checkout context so the original doesn't make any sense.

File: features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature (hunk 2 above)
Comment (NoResponseMate) [inline features/checkout/shipping_order/preventing_placing_order_with_disabled_shipping_method.feature]:
```suggestion
    Scenario: Being prevented from placing an order with a shipping method that's disabled after completing the shipping method choice step
```
?

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $value->getOrderTokenValue()]);
+
+        /** @var OrderInterface $order */
+        Assert::isInstanceOf($order, OrderInterface::class);
+
+        foreach ($order->getShipments() as $shipment) {
+            /** @var ShippingMethodInterface $shippingMethod */
+            $shippingMethod = $shipment->getMethod();
+
+            if ($shippingMethod->isEnabled() && $shippingMethod->getChannels()->contains($order->getChannel())) {
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php]: Is the channel case covered in behat?

File: src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml [hunk 5]
```diff
@@ -72,6 +72,7 @@ sylius:
         payment_method_eligibility: 'This payment method %paymentMethodName% has been disabled. Please reselect your payment method.'
         product_eligibility: 'This product %productName% has been disabled.'
         shipping_method_eligibility: 'Product does not fit requirements for %shippingMethodName% shipping method. Please reselect your shipping method.'
+        shipping_method_not_available: 'The "%shippingMethodName%" shipping method is not available, please choose another one.'
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml]:
```suggestion
        shipping_method_not_available: 'The "%shippingMethodName%" shipping method is not available. Please reselect your shipping method.'
```
Same phrasing as above.

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 6]
```diff
@@ -1164,6 +1180,20 @@ public function iShouldBeInformedThatMyPromotionIsNoLongerApplied(PromotionInter
         );
     }
 
+    /**
+     * @Then I should not be able to confirm order because the :shippingMethod shipping method is not available
+     */
+    public function iShouldNotBeAbleToConfirmOrderBecauseTheShippingMethodIsNotAvailable(ShippingMethodInterface $shippingMethod): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]: There is no need to use a transformer for getting the entire entity as the name will be enough 🖖🏻 
```suggestion
     * @Then I should not be able to confirm order because the :shippingMethodName shipping method is not available
     */
    public function iShouldNotBeAbleToConfirmOrderBecauseTheShippingMethodIsNotAvailable(string $shippingMethodName): void
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 7]
```diff
@@ -501,6 +501,22 @@ public function iProceedThroughCheckoutProcess(): void
         $this->iChoosePaymentMethod($paymentMethod);
     }
 
+    /**
+     * @Given I have proceeded through checkout process with :shippingMethod shipping method
+     */
+    public function iProceedThroughCheckoutProcessWithShippingMethod(ShippingMethodInterface $shippingMethod): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]:
```suggestion
    public function iHaveProceededThroughCheckoutProcessWithShippingMethod(ShippingMethodInterface $shippingMethod): void
```

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 8]
```diff
@@ -546,6 +546,18 @@ public function thisShippingMethodIsOnlyAvailableForOrdersWithATotalWeightLessOr
         $this->addRuleToShippingMethod($rule, $shippingMethod);
     }
 
+    /**
+     * @Given /^(this shipping method) has been disabled$/
+     */
+    public function thisShippingMethodHasBeenDisabled(ShippingMethodInterface $shippingMethod): void
+    {
+        /** @var ShippingMethodInterface $shippingMethod */
+        $shippingMethod = $this->shippingMethodRepository->findOneBy(['code' => $shippingMethod->getCode()]);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php:556]: Why is this necessary?

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutCompleteContext.php [hunk 9]
```diff
@@ -369,6 +369,20 @@ public function iShouldNotBeAbleToProceedCheckoutCompleteStep(): void
         throw new UnexpectedPageException('It should not be possible to complete checkout complete step.');
     }
 
+    /**
+     * @Then I should not be able to confirm order because the :shippingMethod shipping method is not available
+     */
+    public function iShouldNotBeAbleToConfirmOrderBecauseTheShippingMethodIsNotAvailable(ShippingMethodInterface $shippingMethod): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutCompleteContext.php]: Similar as above, there is no need to make a db query to get an entity, as the name will be enough

File: src/Sylius/Behat/Context/Ui/Shop/CheckoutContext.php [hunk 10]
```diff
@@ -98,6 +101,18 @@ public function iProceedThroughCheckoutProcess(string $localeCode = 'en_US', ?st
         $this->paymentContext->iCompleteThePaymentStep();
     }
 
+    /**
+     * @Given I have proceeded through checkout process with :shippingMethod shipping method
+     */
+    public function iProceedThroughCheckoutProcessWithShippingMethod(ShippingMethodInterface $shippingMethod): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/CheckoutContext.php]:
```suggestion
    public function iHaveProceededThroughCheckoutProcessWithShippingMethod(ShippingMethodInterface $shippingMethod): void
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml [hunk 11]
```diff
@@ -41,6 +41,11 @@
                 <value>sylius_checkout_complete</value>
             </option>
         </constraint>
+        <constraint name="Sylius\Bundle\ApiBundle\Validator\Constraints\OrderShippingMethodAvailable">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml]:
```suggestion
        <constraint name="Sylius\Bundle\ApiBundle\Validator\Constraints\OrderShippingMethodAvailability">
```
to be consistent with other constraints

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailable.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class OrderShippingMethodAvailable extends Constraint
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailable.php]:
```suggestion
final class OrderShippingMethodAvailability extends Constraint
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+use Sylius\Bundle\ApiBundle\Command\OrderTokenValueAwareInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+final class OrderShippingMethodAvailableValidator extends ConstraintValidator
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php]: I wonder whether we should not check this as an eligibility checker for the shipping, it makes more sense to me to use this concept. And what is more, it could be used in both contexts, UI and API without necessary of adding duplicated constraints and validators

File: src/Sylius/Behat/Context/Setup/ShippingContext.php (hunk 8 above)
Comment (jakubtobiasz) [reply src/Sylius/Behat/Context/Setup/ShippingContext.php:556]: Yup, for some reason, the ShippingMethod object is not tracked by the entity manager 🤷🏼‍♂️.

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php (hunk 13 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php]: I moved the logic to the Eligibility validator.

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailable.php (hunk 12 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailable.php]: I moved the logic to the Eligibility validator.

File: src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml (hunk 11 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ApiBundle/Resources/config/validation/CompleteOrder.xml]: I moved the logic to the Eligibility validator.

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php (hunk 4 above)
Comment (jakubtobiasz) [reply src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderShippingMethodAvailableValidator.php]: Added

File: src/Sylius/Bundle/CoreBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php [hunk 14]
```diff
@@ -42,9 +43,21 @@ public function validate($value, Constraint $constraint): void
...
+            $shippingMethod = $shipment->getMethod();
+
+            if (!$shippingMethod->isEnabled() || !$shippingMethod->getChannels()->contains($value->getChannel())) {
+                $this->context->addViolation(
+                    $constraint->getMethodNotAvailableMessage(),
+                    ['%shippingMethodName%' => $shippingMethod->getName()],
+                );
+
+                continue;
+            }
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Validator/Constraints/OrderShippingMethodEligibilityValidator.php:56]: I've suggested putting this logic in the new shipping eligibility checker instead of validator 🤔 However, there would be a problem with the message which could be misleading

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Jacob! :1st_place_medal:
